### PR TITLE
Fix #18665, #18559: Add API for getting g2 icons by name

### LINKED ISSUE #18665: Create API to fetch an icon
Something like:
```
let myIcon = getIcon("coin");
```
which returns the proper ID.

### PR DESCRIPTION
**Pull request by spacek531**
This PR attempts to solve the issues created by #18497. It creates an API to get G2 icons by string. It allows plugin creators to use strings to refer to icons by name instead of a hardcoded number. Plugins with old targetAPI will map to the new scheme and plugins with old targetAPI will map back to the old scheme when requesting the icon from the button (just in case).

### CODE REVIEW

File: src/openrct2/scripting/G2Names.hpp [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        { "selection_edge_nw", SPR_G2_SELECTION_EDGE_NW },
+        { "selection_edge_ne", SPR_G2_SELECTION_EDGE_NE },
+        { "selection_edge_sw", SPR_G2_SELECTION_EDGE_SW },
+        { "selection_edge_se", SPR_G2_SELECTION_EDGE_SE },
+    };
+
+    // Maps legacy G2 icons sprite indices to new sprite indices
+
+    static std::map<uint32_t, uint32_t> LegacyIconMap{
+        { 32248, 29357 }, { 29363, 29358 }, { 29357, 29359 }, { 29358, 29360 }, { 29359, 29361 }, { 29360, 29362 },
```
Comment (spacek531) [inline src/openrct2/scripting/G2Names.hpp:118]: I am refactoring this table to use constants like the other one

**Comment by spacek531** [comment]
Does anyone know what's going on with the linux build?

File: src/openrct2/sprites.h [hunk 2]
```diff
@@ -12,6 +12,8 @@
 #include "drawing/ScrollingText.h"
 #include "rct1/Limits.h"
 
+// When adding icon sprites to G2, insert them before the fonts in G2 and give them a name in src/openrct2/scripting/G2Names.hpp
+
```
Comment (karst) [inline src/openrct2/sprites.h:16]: I think it's smarter to place this comment down at the icons at line 885

**Comment by ltsSmitty** [comment]
Since the type in the d.ts isn't an Enum, can you alphabetize the list of icon names for legibility?

File: src/openrct2/sprites.h [hunk 3]
```diff
@@ -881,6 +881,7 @@ enum
     SPR_G2_PLACEHOLDER,
 
     // G2 UI Elements
+    // When UI elements to G2, insert them before fonts and give them a name in src/openrct2/scripting/G2Names.hpp
```
Comment (Basssiiie) [inline src/openrct2/sprites.h:884]: When _adding_ UI elements*

I'm guessing? :)

**Review by Basssiiie — APPROVED** [review]
Looks good to me. Thank you so much for putting your time into this, much appreciated! 🤩

How it looks ingame:
image

**Comment by spacek531** [comment]
I've added a reverse map when getting the image number since it's theoretically possible a plugin would want that.

File: distribution/openrct2.d.ts [hunk 4]
```diff
@@ -2538,7 +2538,7 @@ declare global {
         ButtonWidget | CheckboxWidget | ColourPickerWidget | CustomWidget | DropdownWidget | GroupBoxWidget |
         LabelWidget | ListViewWidget | SpinnerWidget | TextBoxWidget | ViewportWidget;
         
-    type IconName = "empty" | "placeholder" | "logo" | "logo_text" | "fast_forward" | "game_speed_indicator" |
+    enum IconName = "empty" | "placeholder" | "logo" | "logo_text" | "fast_forward" | "game_speed_indicator" |
```
Comment (Basssiiie) [inline distribution/openrct2.d.ts:2541]: This isn't valid a valid Typescript enum, the `type` was correct. @ltsSmitty just asked for the strings to be alphabetical order. :)

File: src/openrct2-ui/scripting/ScWidget.hpp [hunk 5]
```diff
@@ -500,16 +501,28 @@ namespace OpenRCT2::Scripting
...
-                widget->image = value;
+                if (value.type() == DukValue::Type::NUMBER)
+                {
+                    if (GetTargetAPIVersion() <= API_VERSION_63_G2_REORDER)
+                        widget->image = NewIconIndex(value.as_uint());
+                    else
+                        widget->image = value.as_uint();
+                }
+                else
+                    widget->image = GetIconByName(value.as_string());
```
Comment (Gymnasiast) [inline src/openrct2-ui/scripting/ScWidget.hpp:525]: Since the `if` has brackets, please put them around the else block too.

**Comment by Gymnasiast** [comment]
From a quick glance, the implementation looks sound. Would want to test it, and have @IntelOrca do his own review.

File: src/openrct2-ui/scripting/ScWidget.hpp [hunk 6]
```diff
@@ -500,16 +501,30 @@ namespace OpenRCT2::Scripting
...
-        void image_set(uint32_t value)
+        void image_set(DukValue value)
         {
             auto widget = GetWidget();
             if (widget != nullptr && widget->type == WindowWidgetType::FlatBtn)
             {
-                widget->image = value;
+                if (value.type() == DukValue::Type::NUMBER)
+                {
+                    if (GetTargetAPIVersion() <= API_VERSION_63_G2_REORDER)
```
Comment (Gymnasiast) [inline src/openrct2-ui/scripting/ScWidget.hpp:519]: Shouldn’t this be

```suggestion
                    if (GetTargetAPIVersion() < API_VERSION_63_G2_REORDER)
```
?

**Comment by IntelOrca** [comment]
Is there a reason to allow the enum for anything other than `getIcon`, and just call `getIcon` when populating the window desc.

**Comment by spacek531** [comment]
It seemed easier to be able to write a string directly to the widget description than require getIcon. @Basssiiie @ltsSmitty you are the primary users, what do you think?

**Comment by Basssiiie** [comment]
I'd say the enum is more convenient. Why add additional steps if it can be done in one step? It keeps it easy. :)

**Comment by spacek531** [comment]
Should a changelog entry be added that the old plugins were fixed? The change occurred in 0.4.3 development.

**Comment by karst** [comment]
I think the best location for it would be under Improved

**Comment by IntelOrca** [comment]
> I'd say the enum as it is currently implemented is more convenient. Why add additional steps if it can be done in one step? It keeps it easy. :)

To reduce maintainance of the code base. Every API that takes an image otherwise needs this special logic.

**Comment by Basssiiie** [comment]
> To reduce maintainance of the code base. Every API that takes an image otherwise needs this special logic.

Fair point, though you could also reduce the CPP side to a one-liner to get the same amount of maintenance reduction, without the extra bloat on the JS side, e.g.
```cpp
widget->image = GetImageFromDukValue(value);
```
And do the `value.type() == STRING` and `value.type() == INTEGER` in there.

**Comment by Basssiiie** [comment]
Also note that the current implementation is backwards compatible and forcing usage of `getIcon` is not.

**Comment by spacek531** [comment]
> Also note that the current implementation is backwards compatible and forcing usage of `getIcon` is not.

Does that factor in at all? The system uses the old numbers for targetAPI 63 and older.

**Comment by spacek531** [comment]
I did my best to implement Basssiiie's suggestion. It's late so it needs a second opinion.

**Review by IntelOrca — APPROVED** [review]

**Comment by spacek531** [comment]
I think testing tabs with no sprite defined needs testing. I am unsure if it will make a blank tab properly.

**Comment by spacek531** [comment]
I tested it and it does make blank tabs properly when a tab image is undefined. If you give it a garbage value like `true` it shows the small grass sprite (sprite 0) (this also applies to button widgets - if the image field is `true` it will supersede the text field). I changed the default value from 0 to SPR_G2_EMPTY which should display nicer.

**Comment by spacek531** [comment]
I changed my mind about the default value. Making the sprite invisible when it encounters an unknown value allows people to intentionally use garbage data, and makes it more difficult to debug when it occurs.

File: distribution/changelog.txt [hunk 7]
```diff
@@ -49,6 +50,7 @@
 - Fix: [#18606] JSON objects do not take priority over the DAT files they supersede.
 - Fix: [#18620] [Plugin] Crash when reading widget properties from windows that have both static and tab widgets.
 - Fix: [#18653] Negative ratings multipliers do not appear in Vehicle tab.
+- Fix: [#18675] [Plugin] Plugins get incorrect g2 image icons.
```
Comment (Gymnasiast) [inline distribution/changelog.txt:53]:
```suggestion
```

The issue was introduced after the last release, so shouldn’t be in the changelog.

File: src/openrct2-ui/scripting/CustomWindow.cpp [hunk 8]
```diff
@@ -228,6 +221,12 @@ namespace OpenRCT2::Ui::Windows
                     result.offset = { AsOrDefault(dukCoord["x"], 0), AsOrDefault(dukCoord["y"], 0) };
                 }
             }
+            else if (dukImage.type() == DukValue::Type::STRING || dukImage.type() == DukValue::Type::NUMBER)
```
Comment (Gymnasiast) [inline src/openrct2-ui/scripting/CustomWindow.cpp:224]: You swapped the if and else around, which doesn’t change how it works but causes an unnecessarily large diff. Please change the order back.

File: src/openrct2/scripting/G2Names.hpp [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+        { "rct1_simulate_on_pressed", SPR_G2_RCT1_SIMULATE_BUTTON_3 },
+        { "normal_selection_6x6", SPR_G2_LAND_TOOL_SIZE_6 },
+        { "mountain_tool_even", SPR_G2_MOUNTAIN_TOOL_EVEN },
+        { "mountain_tool_odd", SPR_G2_MOUNTAIN_TOOL_ODD },
+        { "scenery_scatter_low", SPR_G2_SCENERY_SCATTER_LOW },
+        { "scenery_scatter_medium", SPR_G2_SCENERY_SCATTER_MEDIUM },
+        { "scenery_scatter_high", SPR_G2_SCENERY_SCATTER_HIGH },
+        { "view", SPR_G2_VIEW },
+        { "path_railings", SPR_G2_PATH_RAILINGS_TAB },
+        { "legacy_path", SPR_G2_LEGACY_PATH_TAB },
```
Comment (Gymnasiast) [inline src/openrct2/scripting/G2Names.hpp:95]:
```suggestion
        { "legacy_paths", SPR_G2_LEGACY_PATH_TAB },
```

File: src/openrct2/scripting/G2Names.hpp [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+        { "path_surface", SPR_G2_PATH_SURFACE_TAB },
+        { "ride_station", SPR_G2_RIDE_STATION_TAB },
+        { "terrain_edge", SPR_G2_TERRAIN_EDGE_TAB },
+        { "hide_vegitation", SPR_G2_BUTTON_HIDE_VEGETATION },
```
Comment (Gymnasiast) [inline src/openrct2/scripting/G2Names.hpp:99]:
```suggestion
        { "hide_vegetation", SPR_G2_BUTTON_HIDE_VEGETATION },
```

Comment (Gymnasiast) [reply src/openrct2/scripting/G2Names.hpp:99]: Don’t forget to also update these two names in the .d.ts file.

**Review by Gymnasiast — APPROVED** [review]
Re-requested a review from @Basssiiie to be sure, but looks good to me now.

File: src/openrct2/scripting/G2Names.hpp [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+        { "zoom_out_background", SPR_G2_ZOOM_OUT_DISABLED },
+        { "map_gen_trees", SPR_G2_TAB_TREE },
+        { "map_gen_noise", SPR_G2_TAB_PENCIL },
+        { "large_scenery", SPR_G2_BUTTON_LARGE_SCENERY },
+        { "small_scenery", SPR_G2_BUTTON_TREES },
+        { "paths", SPR_G2_BUTTON_FOOTPATH },
+        { "rct1_close_off", SPR_G2_RCT1_CLOSE_BUTTON_0 },
+        { "rct1_close_off_pressed", SPR_G2_RCT1_CLOSE_BUTTON_1 },
+        { "rct1_close_on", SPR_G2_RCT1_CLOSE_BUTTON_2 },
+        { "rct1_close_off_pressed", SPR_G2_RCT1_CLOSE_BUTTON_3 },
```
Comment (Basssiiie) [inline src/openrct2/scripting/G2Names.hpp:43]: Should be `"rct1_close_on_pressed"` instead of "off".

File: src/openrct2/scripting/G2Names.hpp [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+        { "rct1_test_off", SPR_G2_RCT1_TEST_BUTTON_0 },
+        { "rct1_test_off_pressed", SPR_G2_RCT1_TEST_BUTTON_1 },
+        { "rct1_test_on", SPR_G2_RCT1_TEST_BUTTON_2 },
+        { "rct1_test_off_pressed", SPR_G2_RCT1_TEST_BUTTON_3 },
```
Comment (Basssiiie) [inline src/openrct2/scripting/G2Names.hpp:47]: Should be `"rct1_test_on_pressed"` instead of "off".

File: src/openrct2/scripting/G2Names.hpp [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+        { 29388, SPR_G2_TITLE_SKIP },
+        { 29389, SPR_G2_SANDBOX },
+        { 29414, SPR_G2_TAB_NEWS },
+        { 29415, SPR_G2_LOCKED },
+        { 29416, SPR_G2_MENU_MULTIPLAYER },
+        { 29433, SPR_G2_SORT },
+        { 29434, SPR_G2_COPY },
+        { 29435, SPR_G2_PASTE },
+        { 29442, SPR_G2_TOOLBAR_MUTE },
+        { 29442, SPR_G2_TOOLBAR_MUTE_PRESSED },
```
Comment (Basssiiie) [inline src/openrct2/scripting/G2Names.hpp:162]: Should be `29443` instead of `29442` (which is MUTE above it).

File: distribution/openrct2.d.ts [hunk 14]
```diff
@@ -2537,6 +2543,25 @@ declare global {
...
+    type IconName = "arrow_down" | "arrow_up" | "chat" | "cheats" | "copy" | "empty" | "eyedropper" |
+        "fast_forward" | "game_speed_indicator" | "game_speed_indicator_double" | "glassy_recolourable" |
+        "hide_full" | "hide_partial" | "hide_scenery" | "hide_supports" | "hide_vegetation" | "hide_vehicles" |
+        "large_scenery" | "legacy_paths" | "link_chain" | "logo" | "logo_text" | "map_east" |
+        "map_east_pressed" | "map_gen_land" | "map_gen_noise" | "map_gen_trees" | "map_north" |
+        "map_north_pressed" | "map_south" | "map_south_pressed" | "map_west" | "map_west_pressed" |
+        "mountain_tool_even" | "mountain_tool_odd" | "multiplayer" | "multiplayer_desync" | "multiplayer_sync" |
+        "multiplayer_toolbar" | "multiplayer_toolbar_pressed" | "mute" | "mute_pressed" | "news_messages" |
+        "normal_selection_6x6" | "paste" | "path_railings" | "path_surface" | "paths" | "placeholder" |
+        "rct1_close_off" | "rct1_close_off_pressed" | "rct1_close_on" | "rct1_open_off" | "rct1_open_off_pressed" |
```
Comment (Basssiiie) [inline distribution/openrct2.d.ts:2556]: `"rct1_close_on_pressed"` and `"rct1_test_on_pressed"` are missing in this string union.

**Review by Basssiiie — COMMENTED** [review]
Accessing sprites by G2 name:
Six Flags 2022-12-10 00-09-08

Accessing by legacy id for old plugins:
Six Flags 2022-12-10 00-09-10

Accessing via getIcon() on newer plugins:
Six Flags 2022-12-10 00-09-22
<details>
  <summary>Code for test plugin:</summary>

```js
const icons = [
	[ "empty",  32248 ],
	[ "placeholder",  29363 ],
	[ "logo",  29357 ],
	[ "logo_text",  29358 ],
	[ "fast_forward",  29359 ],
	[ "game_speed_indicator",  29360 ],
	[ "game_speed_indicator_double",  29361 ],
	[ "map_gen_land",  29362 ],
	[ "zoom_in",  29364 ],
	[ "zoom_in_background",  29365 ],
	[ "zoom_out",  29366 ],
	[ "zoom_out_background",  29367 ],
	[ "map_gen_trees",  29368 ],
	[ "map_gen_noise",  29369 ],
	[ "large_scenery",  29370 ],
	[ "small_scenery",  29371 ],
	[ "paths",  29372 ],
	[ "rct1_close_off",  29373 ],
	[ "rct1_close_off_pressed",  29374 ],
	[ "rct1_close_on",  29375 ],
	[ "rct1_close_on_pressed",  29376 ],
	[ "rct1_test_off",  29377 ],
	[ "rct1_test_off_pressed",  29378 ],
	[ "rct1_test_on",  29379 ],
	[ "rct1_test_on_pressed",  29380 ],
	[ "rct1_open_off",  29381 ],
	[ "rct1_open_off_pressed",  29382 ],
	[ "rct1_open_on",  29383 ],
	[ "rct1_open_on_pressed",  29384 ],
	[ "title_restart",  29385 ],
	[ "title_stop",  29386 ],
	[ "title_play",  29387 ],
	[ "title_skip",  29388 ],
	[ "cheats",  29389 ],
	[ "news_messages",  29414 ],
	[ "server_password",  29415 ],
	[ "multiplayer",  29416 ],
	[ "sort",  29433 ],
	[ "copy",  29434 ],
	[ "paste",  29435 ],
	[ "mute",  29442 ],
	[ "mute_pressed",  29443 ],
	[ "unmute",  29444 ],
	[ "unmute_pressed",  29445 ],
	[ "search",  29461 ],
	[ "eyedropper",  29467 ],
	[ "chat",  29468 ],
	[ "map_north",  29469 ],
	[ "map_north_pressed",  29470 ],
	[ "map_west",  29471 ],
	[ "map_west_pressed",  29472 ],
	[ "map_south",  29473 ],
	[ "map_south_pressed",  29474 ],
	[ "map_east",  29475 ],
	[ "map_east_pressed",  29476 ],
	[ "multiplayer_toolbar",  29477 ],
	[ "multiplayer_toolbar_pressed",  29478 ],
	[ "multiplayer_sync",  29479 ],
	[ "multiplayer_desync",  29480 ],
	[ "simulate",  29481 ],
	[ "rct1_simulate_off",  29482 ],
	[ "rct1_simulate_off_pressed",  29483 ],
	[ "rct1_simulate_on",  29484 ],
	[ "rct1_simulate_on_pressed",  29485 ],
	[ "normal_selection_6x6",  29486 ],
	[ "mountain_tool_even",  29487 ],
	[ "mountain_tool_odd",  29488 ],
	[ "scenery_scatter_low",  29489 ],
	[ "scenery_scatter_medium",  29490 ],
	[ "scenery_scatter_high",  29491 ],
	[ "view",  29494 ],
	[ "path_railings",  29495 ],
	[ "legacy_paths",  29496 ],
	[ "path_surface",  29497 ],
	[ "ride_station",  29498 ],
	[ "terrain_edge",  29499 ],
	[ "hide_vegetation",  29500 ],
	[ "hide_scenery",  29501 ],
	[ "hide_vehicles",  29502 ],
	[ "hide_supports",  29503 ],
	[ "hide_partial",  29504 ],
	[ "hide_full",  29505 ],
	[ "link_chain",  29506 ],
	[ "sideways_tab",  29507 ],
	[ "sideways_tab_active",  29508 ],
	[ "arrow_up",  29509 ],
	[ "arrow_down",  29510 ],
	[ "reload",  29511 ],
].sort(function(a, b) { return a[0].localeCompare(b[0]) });

registerPlugin({
	name: 'Test G2',
	version: '1',
	authors: ['Basssiiie'],
	type: 'remote',
	licence: 'MIT',
	//targetApiVersion: 65,
	main() {
		ui.registerMenuItem("Test G2 icons", function()
		{
			ui.openWindow({
				classification: "g2-test", title: "G2 test",
				x: 20, y: 50, width: 1800, height: 900,
				tabs: [
					{
						image: "zoom_in",
						widgets: create("new g2 name", icons, function(icon) { return icon[0] })
					},
					{
						image: 29499, // terrain edge shovel
						widgets: create("legacy id", icons, function(icon) { return icon[1] })
					},
					{
						image: context.getIcon("chat"),
						widgets: create("getIcon()", icons, function(icon) { return context.getIcon(icon[0]) })
					}
				]
			});
		});
	},
});


function create(name, icons, getId)
{
	var widgets = [
		{ type: "label", text: name, x: 100, y: 30, width: 200, height: 20 },
	];
	var columns = 14
	var span = 130;

	for (var i = 0; i < icons.length; i++)
	{
		var icon = icons[i];
		var id = getId(icon);
		var x = 20 + (i % columns) * span;
		var y = 50 + Math.floor(i / columns) * span;
		widgets.push({
			type: "button", image: id, x: x, y: y, width: 32, height: 32
		});
		widgets.push({
			type: "label", text: (icon[0] + "\n" + icon[1]), x: x, y: y + 40, width: span, height: 32
		});
	}
	return widgets;
}
```

</details>

So besides the few small issues, LGTM! Thank you all for your time.

File: distribution/openrct2.d.ts [hunk 15]
```diff
@@ -2537,6 +2543,26 @@ declare global {
...
+        
+    type IconName = "arrow_down" | "arrow_up" | "chat" | "cheats" | "copy" | "empty" | "eyedropper" |
+        "fast_forward" | "game_speed_indicator" | "game_speed_indicator_double" | "glassy_recolourable" |
+        "hide_full" | "hide_partial" | "hide_scenery" | "hide_supports" | "hide_vegetation" | "hide_vehicles" |
+        "large_scenery" | "legacy_paths" | "link_chain" | "logo" | "logo_text" | "map_east" |
+        "map_east_pressed" | "map_gen_land" | "map_gen_noise" | "map_gen_trees" | "map_north" |
+        "map_north_pressed" | "map_south" | "map_south_pressed" | "map_west" | "map_west_pressed" |
+        "mountain_tool_even" | "mountain_tool_odd" | "multiplayer" | "multiplayer_desync" | "multiplayer_sync" |
+        "multiplayer_toolbar" | "multiplayer_toolbar_pressed" | "mute" | "mute_pressed" | "news_messages" |
+        "normal_selection_6x6" | "paste" | "path_railings" | "path_surface" | "paths" | "placeholder" |
```
Comment (Gymnasiast) [inline distribution/openrct2.d.ts:2555]: path_surface > path_surfaces

**Review by Basssiiie — APPROVED** [review]

File: distribution/openrct2.d.ts (hunk 15 above)
Comment (Gymnasiast) [reply distribution/openrct2.d.ts:2555]: No, they are not. All the tab images for objects should use plural - if these two do not, then their file names should be changed (in another PR).

**Review by Gymnasiast — CHANGES_REQUESTED** [review]
Please fix the icon names as I requested.

Also change `G2Names.hpp` to `IconNames.hpp` - icons from g1 should be added as well (in another PR, preferably).

**Comment by IntelOrca** [comment]
> Also change G2Names.hpp to IconNames.hpp - icons from g1 should be added as well (in another PR, preferably).

Use `IconNames.h`. `.hpp` is for header-only implementations (i.e. fully implemented classes without a .cpp file).

**Comment by Gymnasiast** [comment]
> > Also change G2Names.hpp to IconNames.hpp - icons from g1 should be added as well (in another PR, preferably).
> 
> Use `IconNames.h`. `.hpp` is for header-only implementations (i.e. fully implemented classes or functions without a .cpp file).

There are a few functions defined in that file, so doesn’t .hpp make more sense here?

**Comment by IntelOrca** [comment]
I didn't realise it was in scripting. I originally made most of the scripting code .hpp to reduce code and definition duplicates. But that has since been changed and now scripting files have been changed into .h and .cpp pairs. So I don't know what the plan is now with new stuff.

**Review by Gymnasiast — APPROVED** [review]

**Pull request merged**